jaxopt/_src/broyden.py

```diff
@@ -0,0 +1,382 @@
+                        c_history: Any,
+                        gamma: float = 1.0,
+                        start: int = 0):
+  return inv_jacobian_product(pytree, c_history, d_history, jnp.conjugate(gamma), start)
+
+
+def init_history(pytree, history_size):
+  fun = lambda leaf: jnp.zeros((history_size,) + leaf.shape, dtype=leaf.dtype)
+  return tree_map(fun, pytree)
+
+
+def update_history(history_pytree, new_pytree, last):
+  fun = lambda history_array, new_value: history_array.at[last].set(new_value)
+  return tree_map(fun, history_pytree, new_pytree)
+
+
+class BroydenState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  value: float
+  stepsize: float
+  error: float
+  d_history: Any
+  c_history: Any
+  gamma: jnp.ndarray
+  aux: Optional[Any] = None
+  failed_linesearch: bool = False
+
+
+@dataclass(eq=False)
+class Broyden(base.IterativeSolver):
+  """Broyden solver.
+
+  Attributes:
+    fun: a function of the form ``fun(x, *args, **kwargs)``.
+    has_aux: whether ``fun`` outputs auxiliary data or not.
+      If ``has_aux`` is False, ``fun`` is expected to be
+        scalar-valued.
+      If ``has_aux`` is True, then we have one of the following
+        two cases.
+      At each iteration of the algorithm, the auxiliary outputs are stored
+        in ``state.aux``.
+
+    maxiter: maximum number of Broyden iterations.
+    tol: tolerance of the stopping criterion.
+
+    stepsize: a stepsize to use (if <= 0, use backtracking line search),
+      or a callable specifying the **positive** stepsize to use at each iteration.
+    linesearch: the type of line search to use: for now only "backtracking" for backtracking
+      line search is available.
+    stop_if_linesearch_fails: whether to stop iterations if the line search fails.
+      When True, this matches the behavior of core JAX.
+    maxls: maximum number of iterations to use in the line search.
+    decrease_factor: factor by which to decrease the stepsize during line search
+      (default: 0.8).
+    increase_factor: factor by which to increase the stepsize during line search
+      (default: 1.5).
+    max_stepsize: upper bound on stepsize.
+    min_stepsize: lower bound on stepsize.
+
+    history_size: size of the memory to use.
+    gamma: the initialization of the inverse Jacobian is going to be gamma * I.
+
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto").
+
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
+
+  Reference:
+    Charles G. Broyden.
+    A Class of Methods for Solving Nonlinear Simultaneous Equations.
+    Equation (4.5) (page 581).
+  """
+
+  fun: Callable
```

Comment:
We called it optimality_fun in other root finding classes https://jaxopt.github.io/stable/root_finding.html But now that I think about it, it will probably cause issues if `has_aux=True`...

Reply:
I am not sure I understand the problem with `has_aux=True`. Happy to change to `optimality_fun` anyway.

Reply:
`optimality_fun` needs to return a pytree of the same shape as the parameter pytree, but if `has_aux=True`, `optimality_fun` will return a tuple `(pytree, aux)`... So keeping the `fun` argument is fine. We'll have to deprecate the `optimality_fun` argument in `Bisection` and others.

Reply:
makes sense!